content/chapters/io/lab/content/file-descriptors.md

```diff
@@ -1,233 +1,100 @@
-
-[Quiz 1](../quiz/o-trunc.md)
-
-[Quiz 2](../quiz/fopen-w.md)
-
-### Practice: Close'em All
-
-Just like you use `open()` to create new file descriptors, you can use [`close()`](https://man7.org/linux/man-pages/man2/close.2.html) to destroy them.
-This clears and frees the open file structure to which that entry in the file descriptor table is pointing.
-Use `close()` on the file descriptors you've opened so far in `support/file-descriptors/open_read_write.c`.
+The command above uses `2> /dev/null` to **redirect** `stderr` to `/dev/null`.
+It roughly translates to "make file descriptor `2` refer `/dev/null`".
 
-Note that you don't have to close file descriptors 0, 1 and 2 manually.
-The standard streams are meant to stay alive throughout the lifetime of the process.
-Just like calling `free()` on a `malloc()`-ed pointer, calling `close()` is not really necessary.
-When a process terminates, the OS closes all its file descriptors the same way it frees all its memory.
+This hints at the following corellation between I/O channels and file descriptors:
 
-And keeping this comparison with `malloc()` and `free()`, closing file descriptors is important when they are created inside a loop, as the file descriptor table's size is limited.
-
-## File Handling Conclusion: libc vs syscalls
+- `stdin` (standard input): 0
+- `stdout` (standard output): 1
+- `stderr` (standard error): 2
 
-Up to now, we can draw some parallels between `fopen()` and `open()`.
-While `fopen()` allows the usage of high-level functions such as `fread()` and `fwrite()`, which, among other things, use `while` loops to ensure they always read the required number of bytes, the libc-specific API is not generic enough.
+All processes start with these 3 default file descriptors.
+This is in fact the reason why the calls to `openat()` from above return `3` - it is the first free index in the file descriptor table.
 
-In the following sections, we'll use file descriptors and `read()` and `write()` to interact with some inter-process-communication mechanisms, such as pipes.
+To illustrate all this information, look at the image below.
+All data structures shown below are stored in the Kernel's memory area and are transparent to the user space and are supposed to be accessed only using [`file operations`](file-operations.md).
 
-The table below shows the higher level API provided by libc and the syscalls it relies on.
-As usual, use the `man` pages when in doubt about either of them.
+![File Descriptors](../media/file-descriptors.svg)
 
-|    libc    |  syscall  |
-| :--------: | :-------: |
-| `fopen()`  | `open()`  |
-| `fread()`  | `read()`  |
-| `fwrite()` | `write()` |
-| `fseek()`  | `lseek()` |
-| `fclose()` | `close()` |
+### Practice
 
-So for most equivalents, just remove the leading `f` when moving from the libc function to the underlying syscall.
+Open `support/file-descriptors/print_errors.c`.
+Use `open()` and `close()` to replicate the effect of `2>/dev/null` in code.
```

Comment:
Maybe you can extend this in another practice item to have students implement simple redirection.